### DISCUSSION TITLE
EdgeQL help is needed. "Link filtering" question.

### DISCUSSION THREAD

**Discussion by scorday** [geldata/gel#8052] [Q&A] [answered]
Hello. I understand that my question is kinda basic, but I could not find any help in older discussions.

I have this schema:

`
type User {
}

type Project{
  private_user: User {
      constraint exclusive;
  };
}
`
 Projects can have and have not private_user link. How can I select projects that has .private_user = {} (unset) and .private_user != user1 ?

I tried:
1)  `select Project { * } filter .private_user != (select User filter User.id = <uuid>$arg limit 1) or not exists .private_user;`
2) `select Project { * } filter not exists .private_user or .private_user.id != <uuid>$arg;`
3) `select Project { * } filter .private_user = <User>{};`

I have project with no private_user in db, but every time i get empty results. What's wrong?

**Comment by vpetrovykh** [comment, marked as answer]
Right, so what you want is to use [coalescing comparison](https://docs.edgedb.com/database/stdlib/generic#operator::coaleq).
There are `?=` and `?!=` that do the kind of thing that you want:

Your expression (1) could be re-written like so:
```
select Project { * } filter .private_user.id ?!= <uuid>$arg;
```
Which will get you all the projects that _don't_  match the argument id as their `private_user.id`. And an empty set id will also not match any id you pass, thus all projects without a `private_user` will also be returned by this query.

The reason why this is happening is that most basic operators when dealing with _sets_ are applied to the cross-product of those sets (so all possible pair combinations). So `select {0, 1, 2} + {10, 20}` will get you `{10, 20, 11, 21, 12, 22}`. But that also means that `select {} + {10, 20}` gets you `{}`. The same applies to `=`, `!=` and even `and` and `or`. When comparing potentially empty sets you should use `?=` and `?!=` instead, if you want the `{}` be treated like a valid comparison value. To combine `{}` with other operators of functions you might need `??`, which lets you provide a default value to be used instead of the `{}`. This is because a missing value doesn't always have a natural interpretation in most cases, so the way for you to provide it is via the coalescing operator `??`:
```
_localdev:main> with a := <int64>{} select (a ?? -100_000) < 23;
{true}
_localdev:main> with a := <int64>{} select (a ?? 100_000) < 23;
{false}
```

**Reply by scorday** [reply]
Thanks!!!
